File: src/main/java/io/jjong/algorithm/linked/ReverseRemoveNode.java
package io.jjong.algorithm.linked;

/**
 * create on 2022/12/06. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 *
 * @author Jongsang Han(henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class ReverseRemoveNode {

  public static void main(String[] args) {
    Node<Integer> node1 = new Node<>(1,
        new Node<>(2,
            new Node<>(5,
                new Node<>(9,
                    new Node<>(10)))));

    Node<Integer> result = remove(node1, 3);
    System.out.println(result);
  }


  public static Node<Integer> remove(Node<Integer> node, int removeIdx) {
    Node<Integer> dummy = new Node<>(0, node);
    Node<Integer> first = dummy.next;
    Node<Integer> second = dummy;
    // 두번째 반복자 보다 removeIdx 만큼 앞서게 배치 한다.
    // 5 - 3 = 9, 10
    while (removeIdx-- > 0) {
      first = first.next;
    }

    // 첫번째 반복자 끝에서 K+1번째 노드를 가르키게 된다.
    // 총 2회 반복하므로
    // second 는 5,9,10
    while (first != null) {
      second = second.next;
      first = first.next;
    }
    // 삭제 로직
    // 5를 지우고 9로 덮어 쓴다.
    second.next = second.next.next;

    return dummy.next;
  }
}

File: archive/src/main/kotlin/io/jjong/algorithm/linked/ReverseRemoveNode.kt
package io.jjong.algorithm.linked

/**
 * 끝에서 [removeIdx]번째 노드를 제거한다. 두 포인터(first/second)를 [removeIdx]만큼
 * 벌려두고 first가 끝에 닿으면 second가 삭제 대상 직전을 가리키도록 한다.
 */
object ReverseRemoveNode {
    fun remove(node: Node<Int>?, removeIdx: Int): Node<Int>? {
        val dummy = Node(0, node)
        var first = dummy.next
        var second: Node<Int> = dummy
        var idx = removeIdx

        // 두 번째 반복자보다 removeIdx 만큼 앞서게 배치한다.
        while (idx-- > 0) {
            first = first?.next
        }
        // first가 끝에 닿으면 second는 삭제 대상 직전 노드를 가리킨다.
        while (first != null) {
            second = second.next!!
            first = first.next
        }
        // 삭제 대상을 건너뛰어 다음 노드로 잇는다.
        second.next = second.next?.next
        return dummy.next
    }
}

fun main() {
    val node1 = Node(1, Node(2, Node(5, Node(9, Node(10)))))
    val result = ReverseRemoveNode.remove(node1, 3)
    println(result)
}
